bokchoi/gcp/gcp.py

```diff
@@ -0,0 +1,262 @@
+        self.credentials = self.authorize_client()
+        self.compute = self.get_authorized_compute()
+        self.storage = self.get_authorized_storage()
+
+    def authorize_client(self):
+        """If the environment variable GOOGLE_APPLICATION_CREDENTIALS or If the Google Cloud SDK is
+        installed and has application default credentials set they are loaded and returned."""
+        try:
+            credentials, _ = google.auth.default()
+            return credentials
+        except exceptions.Unauthorized as e:
+            credentials = service_account.Credentials.from_service_account_file(self.gcp.get('AuthKeyLocation'))
+            return credentials
+        except Exception as e:
+            print('Authentication failed, please set the GOOGLE_APPLICATION_CREDENTIALS env variable, install the'
+                  'Google SDK and authenticate, or supply the JSON file location. \n', e)
+
+    def get_authorized_compute(self):
+        """Authorize with default method (implicit env variable) or otherwise
+        use the explicit authentication.
+        """
+        return googleapiclient.discovery.build('compute', 'v1', credentials=self.credentials)
+
+    def get_authorized_storage(self):
+        """Authorize with default method (implicit env variable) or otherwise
+        use the explicit authentication.
+        """
+        return storage.Client(credentials=self.credentials)
+
+    @staticmethod
+    def retrieve_gcp_settings(settings):
+        """
+        Checks the given settings and validates if all required values
+        are there and the values are correct.
+        :arg settings: a json file with with defined settings
+        :return: a python dict with renamed input parameters + defaults
+        """
+        gcp = settings['GCP']
+
+        def check_none(v):
+            if not gcp.get(v):
+                raise Exception('{} is required, please add it to the config'.format(v))
+
+        required = ['ProjectId', 'AuthKeyLocation', 'Bucket']
+        [check_none(v) for v in required]
+
+        return {
+            'project': gcp.get('ProjectId'),
+            'auth_key': gcp.get('AuthKeyLocation'),
+            'bucket': gcp.get('Bucket'),
+            'region': gcp.get('Region', 'europe-west4'),
+            'zone': gcp.get('Zone', 'europe-west4-b'),
+            'network': gcp.get('Network', 'default'),
+            'sub_network': gcp.get('SubNetwork', 'default'),
+            'instance_type': gcp.get('InstanceType', 'n1-standard-1'),
+            'preemptible': gcp.get('Preemptible', False),
+            'disk_space': gcp.get('DiskSpaceGb', 50)
+        }
+
+    def list_instances(self):
+        """List names of all existing instances"""
+        instances = []
+        result = self.compute.instances().list(
+            project=self.gcp.get('project'),
+            zone=self.gcp.get('zone')).execute()
+        [instances.append(x['name']) for x in result['items']]
```

Comment:
I think the list comprehension does not add anything here. I would gor for:
`instances = [x['name'] for x in result['items']]`
and remove `instances = []` above it.

Reply:
good call fixing it!